eve-tools/bpftrace-compiler/edgeview.go

```diff
@@ -0,0 +1,255 @@
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (er *edgeviewRun) stopHTTPDebugPort() {
+	_, err := er.tty.Write([]byte{'\x03'})
+	if err != nil {
+		_, ok := err.(*fs.PathError)
+		if !ok {
+			log.Printf("could not send ctrl+c: %v", err)
+		}
+	}
+	if er.cmd != nil {
+		_ = er.cmd.Process.Signal(os.Interrupt)
+	}
+
+	if er.cancel != nil {
+		er.cancel()
+		er.ctx = nil
+		er.port = 0
+	}
+	if er.cmd != nil {
+		err = er.cmd.Wait()
+		if err != nil {
+			exitErr, ok := err.(*exec.ExitError)
+			if ok {
+				status := exitErr.Sys().(syscall.WaitStatus)
+				if !status.Signaled() {
+					log.Printf("expected that the process exited because of signal, but %v", err)
+				}
+			} else if !strings.Contains(err.Error(), "Wait was already called") {
+				log.Printf("waiting failed: %v/%T", err, err)
+			}
+		}
+	}
+	err = er.tty.Close()
+	if err != nil {
+		_, ok := err.(*fs.PathError)
+		if !ok {
+			log.Printf("closing tty failed: %v", err)
+		}
+	}
+
+	if er.cmd != nil {
+		_ = er.cmd.Process.Kill()
+		er.cmd = nil
+	}
+}
+
+func (er *edgeviewRun) forwardHTTPDebugPort() error {
+	var err error
+
+	er.ctx, er.cancel = context.WithCancel(context.Background())
+	er.cmd = exec.CommandContext(er.ctx, "bash", er.path, "tcp/127.0.0.1:6543")
+
+	er.tty, err = pty.Start(er.cmd)
+	if err != nil {
+		log.Fatalf("starting pty for cmd (%v) failed: %v", er.cmd, err)
+	}
+
+	matchedChan := make(chan struct{})
+
+	var forwardLine string
+	go func() {
+
+		var output string
+		scanner := bufio.NewScanner(er.tty)
+		for scanner.Scan() {
+			line := scanner.Text()
+			output += line + "\n"
+
+			if strings.Contains(line, "127.0.0.1:6543") {
+				forwardLine = strings.TrimFunc(line, func(r rune) bool {
+					return !unicode.IsGraphic(r)
+				})
+				close(matchedChan)
+			}
+		}
+	}()
+
+	select {
+	case <-er.ctx.Done():
+		return nil
+	case <-matchedChan:
+	}
+
+	er.port, err = matchEdgeviewIPOutput(forwardLine)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func matchEdgeviewIPOutput(line string) (uint16, error) {
+	rex := regexp.MustCompile(`(\S+) -> `)
+	matches := rex.FindStringSubmatch(line)
+	if len(matches) != 2 {
+		return 0, fmt.Errorf("could not match '%s'", line)
+	}
+	_, portString, err := net.SplitHostPort(matches[1])
+	if err != nil {
+		return 0, fmt.Errorf("could not extract port from '%s': %v", matches[1], err)
+	}
+
+	port, err := strconv.ParseUint(portString, 10, 16)
+	if err != nil {
+		return 0, fmt.Errorf("could not convert %s to int: %v", portString, err)
+	}
+
+	return uint16(port), nil
+}
+
+func (er edgeviewRun) runEdgeview(edgeviewParam string, matchOutput string) (string, error) {
+	cmd := exec.Command("bash", er.path, edgeviewParam)
+
+	tty, err := pty.Start(cmd)
+	if err != nil {
+		log.Fatalf("starting pty for cmd (%v) failed: %v", cmd, err)
+	}
+
+	defer tty.Close()
+
+	scanner := bufio.NewScanner(tty)
+	success := false
+	var output string
+	for scanner.Scan() {
```

Comment:
Can we add some timeout here?... In case the docker-run is running something unexpected and we are stuck there...

Reply:
You mean a timeout for the `er.cmd`? So when this times out, the pty gets cancelled and with this the scanner gets cancelled?

Reply:
Yep... Not to wait forever if something is broken in the EdgeView container.

Reply:
Something like you do here
https://github.com/lf-edge/eve/blob/41f9de58b6e626bf4cffe6a0b52ccbcf7352abb7/eve-tools/bpftrace-compiler/edgeview.go#L145

Reply:
I would do it like this: https://github.com/lf-edge/eve/pull/4230/files#diff-f06730dae0ea5c70ced78f0970a10f050a2b2da145b322433a8b2b6176ff198aR135

The timeout is generous, but internet connections can be slow.

Reply:
The timeout is 5 minutes on the context; if the context is cancelled:
1. bash is stopped
2. stdout/stderr of bash is closed
3. tty is closed
4. scanner cannot read anymore and returns false

Reply:
I was at the wrong method, now it is there: https://github.com/lf-edge/eve/compare/7dce6732ca22ab10ff90a8843bdc3492a3eccb40..b9853f60e00971e6d685a2abe410f389aa5a6138